add-on/chrome/bootstrap.js

```diff
@@ -346,6 +348,132 @@ var WindowListener = {
+        let ticket = Services.prefs.getIntPref("loop.copy.ticket");
+        if (ticket < 0) {
+          ticket = Math.floor(Math.random() * COPY_MAX_TICKET);
+          Services.prefs.setIntPref("loop.copy.ticket", ticket);
+        }
+        // Short circuit if the special ticket value indicates we're good to go.
+        else if (ticket >= COPY_MAX_TICKET) {
+          this.addCopyPanel();
+          return;
+        }
+
+        // Look up the DNS A-record of a throttler hostname to decide to show.
+        this._dns.asyncResolve(
+          Services.prefs.getCharPref("loop.copy.throttler"),
+          this._dns.RESOLVE_DISABLE_IPV6,
+          (request, record) => {
+            // Use a specific part of the A-record IP address depending on the
+            // channel. I.e., 127.[release/beta/other].[aurora].[nightly].
+            let index = 1;
+            switch (Services.prefs.getCharPref("app.update.channel")) {
+              case "aurora":
+                index = 2;
+                break;
+              case "nightly":
+                index = 3;
+                break;
+            }
+
+            // Check if the 8-bit threshold (0-255) exceeds the ticket (0-254).
+            let threshold = record && record.getNextAddrAsString().split(".")[index];
+            if (threshold && ticket < threshold) {
+              // Remember that we're good to go to avoid future DNS checks.
+              Services.prefs.setIntPref("loop.copy.ticket", COPY_MAX_TICKET);
+              this.addCopyPanel();
+            }
+          },
+          Services.tm.mainThread);
+      },
+
+      /**
+       * Hook into the location bar copy command to open up the copy panel.
+       */
+      addCopyPanel() {
+        // Make a copy of the loop panel as a starting point for the copy panel.
+        let copy = this.panel.cloneNode(false);
+        copy.id = "loop-copy-notification-panel";
+        this.panel.parentNode.appendChild(copy);
+
+        let addTelemetry = bucket => {
+          this.LoopAPI.sendMessageToHandler({
+            data: ["LOOP_COPY_PANEL_ACTIONS", this.constants.COPY_PANEL[bucket]],
+            name: "TelemetryAddValue"
+          });
+        };
+
+        // Override the default behavior of the copy command.
+        let controller = gURLBar._copyCutController;
+        controller._doCommand = controller.doCommand;
+        controller.doCommand = function() {
+          // Do the normal behavior first.
+          this._doCommand.apply(this, arguments);
+
+          // Only ever show the panel once for now.
+          Services.prefs.setBoolPref("loop.copy.shown", true);
+
+          // Open up the copy panel at the loop button.
+          LoopUI.PanelFrame.showPopup(
+            window,
+            LoopUI.toolbarButton.anchor,
+            "loop-copy",
+            null, // Origin
+            "chrome://loop/content/panels/copy.html",
+            null, // Size
+            iframe => {
```

Comment:
Please also break this callback out into its own named function for improved readability of the parent function.